src/Components/Server/src/Circuits/RemoteJSDataStream.cs

```diff
@@ -0,0 +1,173 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.JSInterop;
+
+namespace Microsoft.AspNetCore.Components.Server.Circuits
+{
+    internal sealed class RemoteJSDataStream : Stream
```

Comment:
After reading this a bit further, I have a better understanding of how the "responsiveness" is achieved.

I think we need to validate here that we can't just have unbound memory growth when a client just sends data through the websocket without "waiting". 

My understanding is that the PipeWriter will block after reaching the threshold and apply backpressure because no more messages will be dispatched after the pipe "blocks" until the current message completes. (Other messages can still potentially buffer at the SignalR level (is there a limit to that?)
* If there is a limit, then that's good, although we need to compute what that upper bound is.
* If there's no limit in SignalR, then we shouldn't wait until we block the pipe and we would need to tear down the circuit if the client is not respecting the limits we imposed.

We do have a limit on the maximum number of bytes we are willing to receive (which would prevent us from overflowing with more data than we are willing to accept).

WRT to errors, I think we need to give it a bit more thought. Right now, I believe an error in the upload process will result on the stream producing an error on the user app. I think that's only valid when the error happens for legitimate causes.

In other circumstances (like the client sending too much data or no data, we should just crush the circuit).

In addition to that, I'm wondering what will happen in the presence of errors. For example, if "I'm uploading a large file over a spotty connection", what happens to the "send" from the clients. 

Do we error out and you have to start the upload over? Can we silently miss a chunk of data and make the next chunk available to the consumer before we notice an error?

I think we should consider adding a sequence number to the chunks and keeping track of the number of chunks we are receiving. If the client sends an old chunk, we should stop the circuit. If we somehow receive a chunk after the one we are expecting, we should discard the data and notify the client of the chunk we are expecting to give them a chance of recovering.

Finally, we need to collect all these assumptions, limits, etc. as well as the list of scenarios we have tried for when we do a security review.

Reply:
> My understanding is that the PipeWriter will block after reaching the threshold and apply backpressure because no more messages will be dispatched after the pipe "blocks" until the current message completes. (Other messages can still potentially buffer at the SignalR level (is there a limit to that?)

Agreed this is good to check. AFAIK imposing backpressure here is precisely what `PipeWriter` gives us.

I'm not clear on the mechanics of the backpressure though. Does it somehow communicate back over the TCP channel to make the client stop sending? Or does the server just drop the incoming packets so that the client has to resend? I hope the server isn't somehow buffering the incoming packets indefinitely.

> In addition to that, I'm wondering what will happen in the presence of errors. For example, if "I'm uploading a large file over a spotty connection", what happens to the "send" from the clients. Do we error out and you have to start the upload over? Can we silently miss a chunk of data and make the next chunk available to the consumer before we notice an error?

AFAIK we can't miss a chunk because SignalR guarantees delivery (as long as the connection as a whole isn't dropped). However, if the connection as a whole is dropped, then we do need to make sure this surfaces as an error.

> I think we should consider adding a sequence number to the chunks and keeping track of the number of chunks we are receiving. If the client sends an old chunk, we should stop the circuit. If we somehow receive a chunk after the one we are expecting, we should discard the data and notify the client of the chunk we are expecting to give them a chance of recovering.

Sounds reasonable. My sense is that it's impossible to get out-of-order delivery if:

 * The client is acting in good faith
 * And we don't have any framework bugs
 * And the SignalR connection isn't dropped (maybe this isn't a problem, but I don't know)

However when any of these are false, we want to know about it and be able to diagnose it. So the sequence number idea may help with that.

Reply:
> WRT to errors, I think we need to give it a bit more thought. Right now, I believe an error in the upload process will result on the stream producing an error on the user app. I think that's only valid when the error happens for legitimate causes.
> 
> In other circumstances (like the client sending too much data or no data, we should just crush the circuit).

What's the difference between producing an error:
![image](https://user-images.githubusercontent.com/14852843/122625630-118d3400-d05b-11eb-969f-95d85ba63713.png)
on the user app and "crushing the circuit"? If we make the error fatal:
![image](https://user-images.githubusercontent.com/14852843/122625657-29fd4e80-d05b-11eb-9e1d-81c93c5c8a1e.png)
would that be sufficient?

Reply:
Getting clarification on the expected SignalR behavior. 

Also implemented chunk sequencing to help alleviate some of these concerns. https://github.com/dotnet/aspnetcore/pull/33491/commits/6e09532bf49eed0f404f3d95e83f49f28849b2cb